# K Fold Cross Validation  vey vey imp impp without this no interview

## => ML step
1. Importing libraries
2. Loading the dataset
3. Data preparation
4. Splitting the dataset
5. model creation
# 6. model training  => K fold Cross Validation
7. prediction


- To solve a problem we do have different machine learning algorithm for same problem
- So we need to understand clearly which model is the best to use

## 1 K Folod Cross Valdation
- Cross-validation is a technique it evaluate the model performance


## 2 Ways to training the model
- So far we learning to spilt the data into train and test datasets
- Once the model got trained then we need to test the model

## 3 Limition 
- During testing the model with test dataset, model may fail in few scenarios because model need to face new scenarios right

## 6 K fold cross validation
- Cross - validation is a technique , it evalute the model performance.
- We used to divide 100 sample into fold , each contains 20 samples
- Then now we can start iteration to test the model in different ways

## 7 Iteration 1
- use the first fold to test the model
- use the remanining fold to training

=> first train
- test
- train
- train
- train
- train

## 8 Iteration 2 
- use the first fold to test the model
- use the remanining fold to training

=> secound train
- train
- test
- train
- train
- train

## 9 Last iteration 
- Repeat the process till to the last fold

=> five train
- train
- train
- train
- train
- test

## 10 Average of scores
- The technique is good to calculate the average of all iterartion scores

In [1]:
# K fold cross validation
# LogisticRegression


# step 1 importing libraries
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split


# step 2 loading datasets
digits=load_digits()

# step 3 data perparation
X=digits.data
y=digits.target

# step 4 splitting the dataset
X_train,X_test,y_train,y_test=train_test_split(X,
                                               y,
                                               test_size=0.3)
print("Splitting the data into train and test")

Splitting the data into train and test


In [7]:
# LogisticRegression

# model creation

# step 1 importing libraries
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

# step 2 loading datasets
digits=load_digits()

# step 3 data perparation
X=digits.data
y=digits.target

# step 4 splitting the dataset
X_train,X_test,y_train,y_test=train_test_split(X,
                                               y,
                                               test_size=0.3)
#print("Splitting the data into train and test")

# step 5 model creation 
model1=LogisticRegression(solver="lbfgs", 
                         max_iter=3000)

# step 6 mode training
model.fit(X_train,y_train)



/opt/anaconda3/lib/python3.13/site-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [22]:
# LogisticRegression
# checking the score


# step 1 importing libraries
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

# step 2 loading datasets
digits=load_digits()

# step 3 data perparation
X=digits.data
y=digits.target

# step 4 splitting the dataset
X_train,X_test,y_train,y_test=train_test_split(X,
                                               y,
                                               test_size=0.3)
#print("Splitting the data into train and test")

# step 5 model creation 
model1=LogisticRegression(solver="lbfgs", 
                         max_iter=3000)

# step 6 mode training
model1.fit(X_train,y_train)

# checking the score 
model1.score(X_test,y_test)

0.9574074074074074

In [14]:
# Applying SVM algorthim

# step 1 importing libraries

from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC

# step 2 loading datasets
digits=load_digits()


# step 3 data perparation

X=digits.data
y=digits.target


# step 4 splitting the dataset

X_train,X_test,y_train,y_test=train_test_split(X,
                                               y,
                                               train_size=0.3)

# step 5 model creation 

svm=SVC()
svm.fit(X_train,y_train)


SVC()

In [27]:
# Applying SVM algorthim 
# checking score


# step 1 importing libraries

from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC  # Support Vector Classifier


# step 2 loading datasets

digits=load_digits()

# step 3 data perparation
X=digits.data
y=digits.target


# step 4 splitting the dataset

X_train,X_test,y_train,y_test=train_test_split(X,
                                               y,
                                               train_size=0.3)

# step 5 model creation 

model2=SVC()
model2.fit(X_train,y_train)

print(model2.score(X_test,y_test))

0.9650238473767886


In [25]:
# Applying Random Forest algorithm

# step 1 importing libraries

from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier


# step 2 loading datasets

digits=load_digits()

# step 3 data perparation
X=digits.data
y=digits.target


# step 4 splitting the dataset

X_train,X_test,y_train,y_test=train_test_split(X,
                                               y,
                                               train_size=0.3)

# step 5 model creation 

model3=RandomForestClassifier(n_estimators=40)
model3.fit(X_train,y_train)

print(model3.score(X_test,y_test))

0.9411764705882353


In [28]:
# only one time training 


# train_test_split

# 1 LogisticRegression
# 2 SVC
# 3 Random Forest 

print(model1.score(X_test,y_test))
print(model2.score(X_test,y_test))
print(model3.score(X_test,y_test))


0.9864864864864865
0.9650238473767886
0.9554848966613673


In [43]:
# Logistic regression model performance using cross_val_score 

from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_digits
from sklearn.linear_model import LogisticRegression

digits=load_digits()

model1=LogisticRegression(solver='lbfgs',max_iter=5000)


# step apply  K fold cross validation

scores1=cross_val_score(model1,digits.data,digits.target,cv=3)

print(scores1)

[0.91986644 0.94156928 0.91652755]


In [42]:
# SVM model performance using cross_val_score

from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_digits
from sklearn.svm import SVC

digits=load_digits()

b=SVC()
scores2=cross_val_score(b,digits.data,digits.target,cv=3)
print(scores2)  



[0.96494157 0.97996661 0.96494157]


In [47]:
#  Random Forest  performance using cross_val_score

from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_digits
from sklearn.ensemble import RandomForestClassifier

digits=load_digits()

model=RandomForestClassifier(n_estimators=40) # n_estimatore number dicsion treee
score=cross_val_score(model,digits.data,digits.target,cv=3)

print(score)


[0.93322204 0.95492487 0.92320534]


# 3 times training  at ones


# K fold cross validation

### 1 LogisticRegression  => [0.91986644 0.94156928 0.91652755]
### 2 SVC => [0.96494157 0.97996661 0.96494157]
### 3 Random Forest => [0.93989983 0.95659432 0.92487479]





In [57]:
# K fold cross validation 3 in one

from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_digits
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC 
from sklearn.ensemble import RandomForestClassifier


digits=load_digits()

model1=LogisticRegression(solver="lbfgs",max_iter=5000)
model2=SVC()
model3=RandomForestClassifier(n_estimators=40)


scorc1=cross_val_score(model1,digits.data,digits.target,cv=3)
scorc2=cross_val_score(model2,digits.data,digits.target,cv=3)
scorc3=cross_val_score(model3,digits.data,digits.target,cv=3)


print("Checking all score")
print()
print("1 LogisticRegression ")
print(scorc1)
print()
print("2 SVM")
print(scorc2)
print()
print("3 Random Forest")
print(scorc3)





Checking all score

1 LogisticRegression 
[0.91986644 0.94156928 0.91652755]

2 SVM
[0.96494157 0.97996661 0.96494157]

3 Random Forest
[0.92487479 0.93989983 0.91819699]


In [61]:
# K fold cross validation 3 in one

import numpy as np
from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_digits
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC 
from sklearn.ensemble import RandomForestClassifier


digits=load_digits()

model1=LogisticRegression(solver="lbfgs",max_iter=5000)
model2=SVC()
model3=RandomForestClassifier(n_estimators=40)


scorc1=cross_val_score(model1,digits.data,digits.target,cv=3)
scorc2=cross_val_score(model2,digits.data,digits.target,cv=3)
scorc3=cross_val_score(model3,digits.data,digits.target,cv=3)




print("average all score")
print()
print("1 LogisticRegression ")
print(np.average(scorc1)*100)
print
print()
print("2 SVM")
print(np.average(scorc2)*100)

print()
print("3 Random Forest")
print(np.average(scorc3)*100)




average all score

1 LogisticRegression 
92.59877573734

2 SVM
96.99499165275459

3 Random Forest
92.87701725097385
